### PR TITLE
Adding progress bar to uploads (SCP-3716)

### PR DESCRIPTION
**Pull request by devonbush**
Adds a progress bar to uploads.  This required converting the api calls to XMLHttpRequests as 'fetch' doesn't support progress notifications

image

I'm sure there's a lot of refinement that could be done with the display, but this is intended to be a 'good-enough-for-now' bar

TO TEST:
1. load the upload wizard with react_upload_wizard feature flag on
2. upload a large file
3. observe that the progress bar displays

### CODE REVIEW

**Review by eweitz — CHANGES_REQUESTED** [review]
Given the huge PR but modest described change, it seems `development` shouldn't be the base branch these changes get merged into.

**Comment by devonbush** [comment]
heh, yup, thanks

**Review by eweitz — APPROVED** [review]

File: app/controllers/api/v1/study_files_controller.rb [hunk 1]
```diff
@@ -361,13 +361,13 @@ def chunk
...
-            render json: {errors: 'create/update should be used for uploading the first chunk'}, status: :unprocessable_entity
+            render json: {errors: 'create/update should be used for uploading the first chunk'}, status: :unprocessable_entity and return
           end
           if content_range[:first_byte] != @study_file.upload_file_size
-            render json: {errors: "Incorrect chunk sent: expected bytes starting with #{@study_file.upload_file_size}, received #{content_range[:first_byte]}" }, status: :unprocessable_entity
+            render json: {errors: "Incorrect chunk sent: expected bytes starting with #{@study_file.upload_file_size}, received #{content_range[:first_byte]}" }, status: :unprocessable_entity and return
           end
         else
-          render json: {errors: 'Missing Content-Range header'}, status: :unprocessable_entity
+          render json: {errors: 'Missing Content-Range header'}, status: :unprocessable_entity and return
```
Comment (bistline) [inline app/controllers/api/v1/study_files_controller.rb:370]: Tiny nit - after doing a quick search, it seems that `:bad_request` is the (somewhat) agreed upon response for a missing & required header.

From: https://stackoverflow.com/questions/10576898/what-is-the-most-appropriate-http-status-code-to-return-if-a-required-header-is

File: app/javascript/lib/scp-api.js [hunk 2]
```diff
@@ -767,3 +769,38 @@ export default async function scpApi(
   }
   throw new Error(response)
 }
+
+/**
+  * similar functionality to scpApi, but uses XMLHttpRequest to enable support for progress events.
```
Comment (bistline) [inline app/javascript/lib/scp-api.js:774]: Great comment - much appreciated.

**Review by bistline — APPROVED** [review]

**Pull request merged**